File: config/src/test/java/com/yahoo/vespa/config/protocol/TraceTest.java
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.vespa.config.protocol;

import com.yahoo.slime.Slime;
import org.junit.Test;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

/**
 * @author Ulf Lilleengen
 */
public class TraceTest {

    @Test
    public void trace_level_limits_tracing() {
        Trace trace = Trace.createNew(3);
        assertTrue(trace.shouldTrace(0));
        assertTrue(trace.shouldTrace(1));
        assertTrue(trace.shouldTrace(2));
        assertTrue(trace.shouldTrace(3));
        assertFalse(trace.shouldTrace(4));
        assertFalse(trace.shouldTrace(5));
        trace.trace(1, "foo");
        trace.trace(1, "foo2");
        trace.trace(2, "bar");
        trace.trace(3, "baz");
        trace.trace(3, "baz2");
        trace.trace(4, "quux");
        trace.trace(5, "quux2");
        String str = trace.toString();
        assertTrue(str.contains("foo"));
        assertTrue(str.contains("foo2"));
        assertTrue(str.contains("bar"));
        assertTrue(str.contains("baz"));
        assertTrue(str.contains("baz2"));
        assertFalse(str.contains("quux"));
        assertFalse(str.contains("quux2"));
    }

    @Test
    public void trace_serialization() {
        Trace trace = Trace.createNew(1);
        trace.trace(0, "foobar");
        trace.trace(1, "barbaz");
        Slime slime = new Slime();
        trace.serialize(slime.setObject());
        Trace trace2 = Trace.fromSlime(slime.get());
        trace2.trace(1, "quux");
        String trace1Str = trace.toString();
        String trace2Str = trace2.toString();
        assertTrue(trace1Str.contains("foobar"));
        assertTrue(trace1Str.contains("barbaz"));
        assertFalse(trace1Str.contains("quux"));

        assertTrue(trace2Str.contains("foobar"));
        assertTrue(trace2Str.contains("barbaz"));
        assertTrue(trace2Str.contains("quux"));
    }

}

File: config/src/tests/trace/trace_test.cpp
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include <vespa/config/common/trace.h>
#include <vespa/vespalib/data/slime/slime.h>
#include <vespa/vespalib/gtest/gtest.h>


using namespace config;
using namespace vespalib;
using namespace vespalib::slime;

struct FixedClock : public Clock
{
    FixedClock() : _currentTime() { }
    ~FixedClock() override;
    vespalib::system_time _currentTime;
    vespalib::system_time currentTime() const override { return _currentTime; }
};

FixedClock::~FixedClock() = default;

TEST(TraceTest, that_trace_can_be_serialized_and_deserialized)
{
    Trace trace(4);
    trace.trace(4, "foo");
    trace.trace(3, "bar");
    trace.trace(5, "baz");

    Slime slime;
    Cursor & cursor(slime.setObject());
    trace.serialize(cursor);

    Trace trace2;
    trace2.deserialize(slime.get());

    Slime slime2;
    trace2.serialize(slime2.setObject());
    Trace trace3;
    trace3.deserialize(slime2.get());

    EXPECT_EQ(trace.toString(), trace3.toString());
}

TEST(TraceTest, that_trace_level_is_taken_into_account)
{
    FixedClock f1;
    f1._currentTime = vespalib::system_time(3ms);
    Trace trace(4, f1);
    trace.trace(4, "foo");
    trace.trace(5, "bar");
    EXPECT_EQ("[\n"
"    {\n"
"        \"timestamp\": 3,\n"
"        \"payload\": \"foo\"\n"
"    }\n"
"]\n", trace.toString());
}

TEST(TraceTest, that_trace_can_be_copied)
{
    Trace trace(3);
    trace.trace(2, "foo");
    trace.trace(3, "bar");
    Trace trace2(trace);
    EXPECT_EQ(trace.toString(), trace2.toString());
}

constexpr vespalib::system_time epoch;

TEST(TraceTest, ensure_that_system_clock_is_used_by_default)
{
    Trace trace(2);
    trace.trace(1, "foo");
    TraceNode child(trace.getRoot().getChild(0));
    EXPECT_TRUE(child.getTimestamp() > epoch);
}

GTEST_MAIN_RUN_ALL_TESTS()
